beacon_node/network/src/subnet_service/mod.rs

```diff
@@ -17,12 +32,640 @@ pub enum SubnetServiceMessage {
+        // The time to the required slot.
+        let time_to_subscription_slot = self
+            .beacon_chain
+            .slot_clock
+            .duration_to_slot(slot)
+            .unwrap_or_default(); // If this is a past slot we will just get a 0 duration.
+
+        // Calculate how long before we need to subscribe to the subnet.
+        let time_to_subscription_start =
+            time_to_subscription_slot.saturating_sub(advance_subscription_duration);
+
+        // The time after a duty slot where we no longer need it in the `aggregate_validators_on_subnet`
+        // delay map.
+        let time_to_unsubscribe =
+            time_to_subscription_slot + UNSUBSCRIBE_AFTER_AGGREGATOR_DUTY * slot_duration;
+        if let Some(tracked_vals) = self.aggregate_validators_on_subnet.as_mut() {
+            tracked_vals.insert_at(ExactSubnet { subnet, slot }, time_to_unsubscribe);
+        }
+
+        // If the subscription should be done in the future, schedule it. Otherwise subscribe
+        // immediately.
+        if time_to_subscription_start.is_zero() {
+            // This is a current or past slot, we subscribe immediately.
+            self.subscribe_to_subnet_immediately(subnet, slot + 1)?;
+        } else {
+            // This is a future slot, schedule subscribing.
+            self.scheduled_subscriptions
+                .insert_at(subnet, time_to_subscription_start);
+        }
+
+        Ok(())
+    }
+
+    /// Adds a subscription event and an associated unsubscription event if required.
+    fn subscribe_to_sync_subnet(
+        &mut self,
+        subnet: Subnet,
+        duration_to_unsubscribe: Duration,
+        slot_required_until: Slot,
+    ) {
+        // Return if we have subscribed to all subnets
+        if self.subscribe_all_subnets {
+            return;
+        }
+
+        // Update the unsubscription duration if we already have a subscription for the subnet
+        if let Some(current_instant_to_unsubscribe) = self.subscriptions.deadline(&subnet) {
+            // The extra 500ms in the comparison accounts of the inaccuracy of the underlying
+            // DelayQueue inside the delaymap struct.
+            let current_duration_to_unsubscribe = (current_instant_to_unsubscribe
+                + Duration::from_millis(500))
+            .checked_duration_since(Instant::now())
+            .unwrap_or(Duration::from_secs(0));
+
+            if duration_to_unsubscribe > current_duration_to_unsubscribe {
+                self.subscriptions
+                    .update_timeout(&subnet, duration_to_unsubscribe);
+            }
+        } else {
+            // We have not subscribed before, so subscribe
+            self.subscriptions
+                .insert_at(subnet, duration_to_unsubscribe);
+            // We are not currently subscribed and have no waiting subscription, create one
+            debug!(self.log, "Subscribing to subnet"; "subnet" => ?subnet, "until" => ?slot_required_until);
```

Comment:
Didn't follow what its not relevant here? I think it is.

Reply:
ah I didn't read the code that was calling this - `duration_to_unsubscribe` was derived from `slot_required_until`.

I initially thought we're not using `slot_required_until` here because it's only used for logging in this function, and we use `duration_to_unsubscribe` when inserting into the `HashSetDelay`.

All good!